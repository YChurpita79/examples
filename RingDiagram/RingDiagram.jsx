import React, { memo } from "react";
import PropTypes from "prop-types";

const RingDiagram = memo(({ config, chartData }) => {
  const offsetRatio = {
    top: 0.25,
    right: 0,
    left: 0.5,
    bottom: -0.25,
  };

  const strokeWidth = config.strokeWidth || 5;
  const radiusValue = config.radius || 100;
  const radius = radiusValue - strokeWidth / 2;
  const fullSize = 2 * radiusValue;
  const length = 2 * Math.PI * radius;
  let startPoint = config.start in offsetRatio ? config.start : "top";
  const chartOffset = length * offsetRatio[startPoint];
  const sectors = [];

  chartData.forEach((sectorData, sectorIndex) => {
    const width = (length * sectorData.value) / 100;
    let offset = chartOffset;

    if (sectorIndex > 0) {
      let prevSector = sectors[sectorIndex - 1];
      offset = prevSector.offset - prevSector.width;
    }

    sectors.push({
      ...sectorData,
      width,
      offset,
    });
  });

  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      version="1.1"
      viewBox={`0 0 ${fullSize} ${fullSize}`}
      fill={"none"}
      width={fullSize}
      height={fullSize}
    >
      {sectors?.map((sector, index) => {
        return (
          <circle
            key={index}
            cx={radius + strokeWidth / 2}
            cy={radius + strokeWidth / 2}
            r={radius}
            strokeDasharray={`${sector.width} ${length - sector.width}`}
            strokeDashoffset={sector.offset}
            stroke={sector.color}
            strokeLinecap={"round"}
            strokeWidth={strokeWidth}
          />
        );
      })}
    </svg>
  );
});

RingDiagram.propTypes = {
  config: PropTypes.object,
  chartData: PropTypes.array,
};
export default RingDiagram;
