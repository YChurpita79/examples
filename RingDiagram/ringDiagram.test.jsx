import React from "react";
import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";

import RingDiagram from "./RingDiagram.jsx";

describe("RingDiagram", () => {
  const chartData = [
    {
      value: 100,
      color: "#D9DFE8",
    },
    {
      value: 50,
      color: "#727272",
    },
  ];

  const config = {
    strokeWidth: 5,
    radius: 24,
    start: "top",
  };

  it("should be rendered", () => {
    const { container } = render(<RingDiagram chartData={chartData} config={config} />);
    expect(container).toMatchSnapshot();
  });
});
