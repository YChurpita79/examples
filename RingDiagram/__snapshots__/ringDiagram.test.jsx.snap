// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RingDiagram > should be rendered 1`] = `
<div>
  <svg
    fill="none"
    height="48"
    version="1.1"
    viewBox="0 0 48 48"
    width="48"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="24"
      cy="24"
      r="21.5"
      stroke="#D9DFE8"
      stroke-dasharray="135.0884841043611 0"
      stroke-dashoffset="33.772121026090275"
      stroke-linecap="round"
      stroke-width="5"
    />
    <circle
      cx="24"
      cy="24"
      r="21.5"
      stroke="#727272"
      stroke-dasharray="67.54424205218055 67.54424205218055"
      stroke-dashoffset="-101.31636307827083"
      stroke-linecap="round"
      stroke-width="5"
    />
  </svg>
</div>
`;
